### ISSUE TITLE
Undefined call to Equals object in collection during flush just before commit 

### ISSUE THREAD

**Issue by eiffelAqua** [nhibernate/nhibernate-core#2141]
Since my upgrade from NH3.4 to 5.2.5, I discover a strange behavior.
Just before the commit on a my collection of entities, a strange call to the override Equals method is fired for each saved entities with empty object. This causes a crash (cast exception) because i don't check the type of object passed in argument for comparison.  
I tried to save my entities with a specific transaction and commit between each object (and clear NH session at each time) and it's ok in this case !

A workaround is to check the typeof of my object in my override Equals method but i need to recode all my entities (more than 400 entity). it's not normal.

I download source code for NH5.2.5 and this produce when the Flush() method is called in FlushBeforeTransactionCompletion method for auto FlushMode.  In this case, the  flushEventListener array contains something...
I can disable the autoFlush mode but it's not recommanded.

Help, please. Thanks

**Comment by gliljas** [comment]
While there may be caude to check the changed behavior, a type check in Equals is "normal" and expected by the method contract. It's usually combined with the null check.

```
if (!(obj is TheType))
{
   return false;
}
```

Perhaps you can automate the modification. It's a perfect fit for a Roslyn codefix, but if your filenames are in order (i.e named as the classes), you could just cycle through them with Powershell or a small C# script (LinqPad?).

**Comment by hazzik** [comment]
@eiffelAqua what is the stack trace?

**Comment by eiffelAqua** [comment]
Thanks,

Bellow the stack trace.

Impossible d'effectuer un cast d'un objet de type 'Metier_AquaTIC.sandre.par.Support' en type 'Metier_AquaTIC.sandre.par.FractionAnalysee'.
à Metier_AquaTIC.sandre.par.FractionAnalysee.Equals(Object obj) dans D:\dev\AquaTIC\trunk\application\Metier_AquaTIC\sandre\par\FractionAnalysee.cs:ligne 56
à System.Collections.Generic.ObjectEqualityComparer`1.Equals(T x, T y)
à System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
à System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
à NHibernate.Engine.ActionQueue.InsertActionSorter.RequireNewBatch(AbstractEntityInsertAction action, Int32 latestBatchNumberForType)
à NHibernate.Engine.ActionQueue.InsertActionSorter.GetBatchNumber(AbstractEntityInsertAction action, String entityName)
à NHibernate.Engine.ActionQueue.InsertActionSorter.Sort()
à NHibernate.Engine.ActionQueue.SortInsertActions()
à NHibernate.Engine.ActionQueue.SortActions()
à NHibernate.Event.Default.AbstractFlushingEventListener.FlushEntities(FlushEvent event)
à NHibernate.Event.Default.AbstractFlushingEventListener.FlushEverythingToExecutions(FlushEvent event)
à NHibernate.Event.Default.DefaultFlushEventListener.OnFlush(FlushEvent event)
à NHibernate.Impl.SessionImpl.Flush()
à NHibernate.Impl.SessionImpl.FlushBeforeTransactionCompletion()
à NHibernate.Impl.SessionImpl.BeforeTransactionCompletion(ITransaction tx)
à NHibernate.Transaction.AdoTransaction.Commit()
à MappingOR_AquaTIC.utils.AquaMappingORHelper.Execute[T](ProcessAction`1 action, T parameter, ProcessAction`1 onSuccessAction, ProcessAction_OnError onErrorAction, Boolean clearSession) dans D:\dev\AquaTIC\trunk\application\MappingOR_AquaTIC\utils\AquaMappingORHelper.cs:ligne 130

Here the code of my business entity 'FractionAnalysee' which i try to save.

using System;
using Metier_AquaTIC.utils;
using Outils_AquaTIC.Helpers;

namespace Metier_AquaTIC.sandre.par
{
    [Serializable]
    public class FractionAnalysee : Entity<FractionAnalysee>, IEquatable<FractionAnalysee>, IComparable<FractionAnalysee>
    {
        #region Accesseurs
        public virtual int IdFractionAnalysee { get { return base.Id; } set { base.Id = value; } }
        public virtual string CdFractionAnalysee { get; set; }
        public virtual string LbFractionAnalysee { get; set; }
        public virtual EnumStDonneeReferentiel StFractionAnalysee { get; set; }
        public virtual DateTime? DateCreationFractionAnalysee { get; set; }
        public virtual DateTime? DateMajFractionAnalysee { get; set; }
        public virtual string AuteurFractionAnalysee { get; set; }
        public virtual string ComFractionAnalysee { get; set; }
        public virtual string NomIntFractionAnalysee { get; set; }
        //[NonSerialized]
        public virtual NatureFractionAnalysee NatureFractionAnalysee { get; set; }
        public virtual Support Support { get; set; }
        public virtual String LbCdFractionAnalysee
        {
            get
            {
                return LbFractionAnalysee + " (" + CdFractionAnalysee + ")";
            }

        }
        #endregion

        #region Constructeurs
        public FractionAnalysee() { }

        public FractionAnalysee(string cdFractionAnalysee)
        {
            this.CdFractionAnalysee = cdFractionAnalysee;
        }
        #endregion

        #region Méthodes
        public override string ToString()
        {
            return "Code fraction :" + CdFractionAnalysee + " libelle :" + LbFractionAnalysee + " statut :" + StFractionAnalysee;
        }

        public virtual string GetKey()
        {
            return CdFractionAnalysee;
        }

        public override bool Equals(object obj)
        {
            bool equal = false;
            equal = Equals((FractionAnalysee)obj);

            /*if (obj is FractionAnalysee)
            {
                equal = Equals((FractionAnalysee)obj);
            }
            else
            {
                int a = 5;
                Console.WriteLine("obj : ", obj.GetType());
                Console.WriteLine("obj toString: ", obj.ToString());
                Console.WriteLine("this : ", this.ToString());
            }
            */
            return equal;

        }

        public virtual bool Equals(FractionAnalysee obj)
        {
            bool objEqual = false;

            bool? entityEqual = base.IsEntityEqual(obj);
            // les 2 objets sont persistants, ils sont égaux si leurs id en base sont identiques
            if (entityEqual.HasValue)
                objEqual = entityEqual.Value;
            else
            {
                // l'un des 2 objets (ou les 2) ne sont pas persistants, on doit effectuer une comparaison "métier" basée sur la clé unique de l'objet afin de déterminer si les 2 objets sont égaux
                string str1 = this.GetKey();
                string str2 = obj.GetKey();
                objEqual = str1.Equals(str2);
            }

            return objEqual;
        }

        public override int GetHashCode()
        {
            int hashCode;

            int? entityHashCode = base.GetEntityHashCode();
            // l'objet est persistant, on peut calculer son HashCode à partir de son id, ce qui évite de passer par Equals
            if (entityHashCode.HasValue)
                hashCode = entityHashCode.Value;
            // l'objet est transient, le HashCode doit être calculé à partir de la clé métier (ou en partie)
            else
                hashCode = GetKey().GetHashCode();

            return hashCode;
        }

        public virtual int CompareTo(FractionAnalysee obj)
        {
            if (obj != null)
                return (LbFractionAnalysee.CompareTo(obj.LbFractionAnalysee));
            else
                return -1;
        }

        public virtual FractionAnalysee This()
        {
            return this;
        }

        public virtual string LibelleCodeFractionAnalysee
        {
            get
            {
                return LbFractionAnalysee + " (" + CdFractionAnalysee + ")";
            }
        }

        public virtual void Validate()
        {
            Anomalie anomalie = new Anomalie(GetType(), "Invalide fraction analysée n° : " + IdFractionAnalysee);

            if (CdFractionAnalysee == null)
                anomalie.Errors.Add("Le code de la fraction analysée est obligatoire");
            else
            {
                try
                {
                    StringExtension.ValideChaine(CdFractionAnalysee, "le code de la fraction analysée");
                }
                catch (StringInvalidException e)
                {
                    anomalie.Errors.Add(e.Message);
                }
            }

            try
            {
                StringExtension.ValideChaine(LbFractionAnalysee, "le libellé de la fraction analysée");
            }
            catch (StringInvalidException e)
            {
                anomalie.Errors.Add(e.Message);
            }

            try
            {
                StringExtension.ValideChaine(AuteurFractionAnalysee, "l'auteur de la fraction analysée");
            }
            catch (StringInvalidException e)
            {
                anomalie.Errors.Add(e.Message);
            }

            try
            {
                StringExtension.ValideChaine(ComFractionAnalysee, "le commentaire de la fraction analysée");
            }
            catch (StringInvalidException e)
            {
                anomalie.Errors.Add(e.Message);
            }

            try
            {
                StringExtension.ValideChaine(NomIntFractionAnalysee, "le nom internationnal de la fraction analysée");
            }
            catch (StringInvalidException e)
            {
                anomalie.Errors.Add(e.Message);
            }

            if (NatureFractionAnalysee != null)
                NatureFractionAnalysee.Validate();

            if (Support != null)
                Support.Validate();

            if (anomalie.HasError)
                throw anomalie;
        }

        public virtual void Copy(FractionAnalysee fracAna)
        {
            //recopie les propriétés de l'objet métier sauf ceux de la clé unique
            this.LbFractionAnalysee = fracAna.LbFractionAnalysee;
            this.StFractionAnalysee = fracAna.StFractionAnalysee;
            this.DateCreationFractionAnalysee = fracAna.DateCreationFractionAnalysee;
            this.DateMajFractionAnalysee = fracAna.DateMajFractionAnalysee;
            this.AuteurFractionAnalysee = fracAna.AuteurFractionAnalysee;
            this.ComFractionAnalysee = fracAna.ComFractionAnalysee;
            this.NomIntFractionAnalysee = fracAna.NomIntFractionAnalysee;
            this.NatureFractionAnalysee = fracAna.NatureFractionAnalysee;
            this.Support = fracAna.Support;
        }
        #endregion
    }

    public class InvalidNatureFractionAnalyseeException : ApplicationException
    {
        #region Attributs
        private FractionAnalysee _f;
        #endregion

        #region Constructeurs
        public InvalidNatureFractionAnalyseeException(FractionAnalysee f)
        {
            _f = f;
        }
        #endregion

        #region Méthodes
        public override string ToString()
        {
            return "Nature de fraction analysée obligatoire pour la faction analysée suivante => " +
                                            "Code fraction analysée : " + _f.CdFractionAnalysee;
        }
        #endregion
    }
}

and the base Entity class

namespace Metier_AquaTIC.utils
{
    public abstract class Entity<TId>
    {
        protected int Id;

        public static bool IsTransient(Entity<TId> obj)
        {
            return obj.Id < 1;//Equals(obj.Id, default(int));//Cas particulier des obsHydro calculées non enregistrées qui ont un id à -1 pour une question d'affichage
        }

        public virtual bool? IsEntityEqual(Entity<TId> other)
        {
            bool? entityEqual = null;

            // les adresses mémoire sont identiques => les objets sont égaux
            if (ReferenceEquals(this, other))
                entityEqual = true;
            // les types des 2 objets sont identiques => => les objets sont égaux. A noter que le code permet de prendre en compte les types proxy.
            else if (GetType().IsAssignableFrom(other.GetType()) || other.GetType().IsAssignableFrom(GetType()))
            {
                // les 2 objets sont persistants et ont le même Id en base => les objets sont égaux
                if (this != null && !IsTransient(this) && other != null && !IsTransient(other))
                    entityEqual = Equals(this.Id, other.Id) ? true : false;
            }

            return entityEqual;
        }

        public virtual int? GetEntityHashCode()
        {
            //Code mis en commentaire : les contains ne fonctionnaient plus dans le cas de recherche d'objet non transcient avec une liste contenant des objets transcients (Synchroniser,...)
            //int? entityHashCode = null;

            //if (!IsTransient(this))
            //    entityHashCode = Id.GetHashCode();

            //return entityHashCode;
            return null;
        }
    }
}


I continue the NH debugging and the problems occurs during the source.ActionQueue.SortActions() in the FlushEntities(@event) method in AbstractFlushingEventListener.cs file

it seems a confusion between <Support> entity and <FractionAnalysee> entity in the RequireNewBatch method because the Equals is called on bad entity during the following code in ActionQueue.cs.
here is the fraction of code that generate the bad Equals call in the RequireNewBatch method in ActionQueue.cs (line 713)

if (_entityBatchNumber.TryGetValue(value, out associationBatchNumber) &&
associationBatchNumber > latestBatchNumberForType)
{
	return true;
}

Important : Note the <Support> entity is referenced inside the <FractionAnalysee> entity.

Here is the stack trace of my debugging session : 
NHibernate.dll!NHibernate.Event.Default.AbstractFlushingEventListener.FlushEverythingToExecutions(NHibernate.Event.FlushEvent event) Ligne 73 	NHibernate.dll!NHibernate.Event.Default.DefaultFlushEventListener.OnFlush(NHibernate.Event.FlushEvent event) Ligne 18
NHibernate.dll!NHibernate.Impl.SessionImpl.Flush() Ligne 1333
NHibernate.dll!NHibernate.Impl.SessionImpl.FlushBeforeTransactionCompletion() Ligne 2022
NHibernate.dll!NHibernate.Impl.SessionImpl.BeforeTransactionCompletion(NHibernate.ITransaction tx) Ligne 2004
NHibernate.dll!NHibernate.Transaction.AdoTransaction.Commit() Ligne 203
MappingOR_AquaTIC.dll!MappingOR_AquaTIC::utils::AquaMappingORHelper::Execute(MappingOR_AquaTIC::utils::AquaMappingORHelper::ProcessAction<T>^ , T , MappingOR_AquaTIC::utils::AquaMappingORHelper::ProcessAction<T>^ , MappingOR_AquaTIC::utils::AquaMappingORHelper::ProcessAction_OnError^ , bool ) Ligne 130
Service_AquaTIC.dll!Service_AquaTIC::majRefSandre::ServiceInitBDDRefSandre::InitFractionAnalysee(System::String^ ) Ligne 4100
Service_AquaTIC.dll!Service_AquaTIC::majRefSandre::ServiceInitBDDRefSandre::InitFractionAnalysee() Ligne 4075
TestService_AquaTIC.exe!TestService_AquaTIC::TestReferentiel::TesterInitRefSandre() Ligne 97
TestService_AquaTIC.exe!TestService_AquaTIC::Program::Main(array<System::String^>^ ) Ligne 43


I agree with the test : 
if (!(obj is TheType))
{
   return false;
}

but it's not the solution and i think there is a real problem under my test case because in my previous NH version 3.4, there is no problem and the project is used since a few years in intensive production environnement.

Thanks in advance for you help on this nice project.

**Comment by eiffelAqua** [comment]
Hi,

After few hours, i understand my problem,..

It comes from the TryGetValue method in the _entityBatchNumber object dictionnary (in ActionQueue.cs). This dictionnary contains mutlitple entities with different types and the TryGetValue generate Equals call in order to find the good entity.  It generates a class cast exception because i don't test the type in my Equals method. Finally, I'm ok with Gunnar Lijlas and i have to change all my object entities.

However, it's possible to make improvment in NH projet ? I suggest to add a try/catch Exception on the TryGetValue in the RequireNewBatch method ? I changed locally the NH source code (see above) and its'ok now. I think it's preferable because you are not sure of the good implementation of Equals in the various project.

For example, in our company we refuse to used mixed object dictionnray in our project and we think that the equals method should always receive an argument of the right kind.

```
private bool RequireNewBatch(AbstractEntityInsertAction action, int latestBatchNumberForType)
			{
				// This method assumes the original action list is already sorted in order to respect dependencies.
				var propertyValues = action.State;
				var propertyTypes = action.Persister.EntityMetamodel?.PropertyTypes;
				if (propertyTypes == null)
				{
					log.Info(
						"Entity {0} persister does not provide meta-data, giving up batching grouping optimization for this entity.",
						action.EntityName);
					// Cancel grouping optimization for this entity.
					return true;
				}

				int latestDependency;
                try
                {
                    if (_entityBatchDependency.TryGetValue(action.Instance, out latestDependency) && latestDependency > latestBatchNumberForType)
                        return true;
                }
                catch (Exception e)
                {
                    log.Info("pb de cast dans la méthode Equals (catch Aquascop n°1)", e.Message);
                }

                for (var i = 0; i < propertyValues.Length; i++)
				{
					var value = propertyValues[i];
					var type = propertyTypes[i];

					if (type.IsEntityType && value != null &&
						// If the value is not initialized, it is a proxy with pending load from database,
						// so it can only be an already persisted entity. It can not have its own pending
						// insertion batch. So there is no need to seek for it, and it avoids initializing
						// it by searching it in a dictionary. Fixes #1338.
						NHibernateUtil.IsInitialized(value))
					{
						// find the batch number associated with the current association, if any.
						int associationBatchNumber;
                        try
                        {
                            if (_entityBatchNumber.TryGetValue(value, out associationBatchNumber) &&
                                associationBatchNumber > latestBatchNumberForType)
                            {
                                return true;
                            }
                        }
                        catch (Exception e)
                        {
                            log.Info("pb de cast dans la méthode Equals (catch Aquascop n°2)", e.Message);
                        }
					}
				}

				return false;
			}

```

**Comment by InlineAsm** [comment]
> we think that the equals method should always receive an argument of the right kind...

It's not right to assume the equals method always receives it's own type of object. If you check different implementations of the equals method you will see that all of them check the object for type equality. Even the microsoft guidelines: https://docs.microsoft.com/en-us/previous-versions/ms173147(v=vs.90) all check for the correct type.

The reason for this check is because the equals method contract specifies it should check if the object given is equal to itself. The object given can be of any type and it's valid to pass any object to the equals method.

>I suggest to add a try/catch Exception on the TryGetValue in the RequireNewBatch method ?
This is a change that would only hide the fact the equals method is wrongly implemented. Hiding programming errors by catching exceptions is a case of 'bad design'
